### PR TITLE
[SPARK-30066][SQL] Support columnar execution on interval types

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Columnar execution support for interval types

### Why are the changes needed?

support cache tables with interval columns
improve performance too

### Does this PR introduce any user-facing change?

Yes cache table with accept interval columns

### How was this patch tested?

add ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc  @cloud-fan @maropu, we may need  https://github.com/apache/spark/pull/26680 to be merged to make this forward here, thanks for reviewing

**Comment by maropu** [comment]
Is the title `Support cache for interval data`?

**Comment by yaooqinn** [comment]
> Is the title `Support cache for interval data`?

cache seems API-like to me, here we also use this cache for underly optimization

**Comment by yaooqinn** [comment]
cc @cloud-fan  @maropu  again, ready for review, thanks for your help.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnAccessor.scala [hunk 1]
```diff
@@ -104,6 +105,10 @@ private[columnar] class BinaryColumnAccessor(buffer: ByteBuffer)
   extends BasicColumnAccessor[Array[Byte]](buffer, BINARY)
   with NullableColumnAccessor
 
+private[columnar] class IntervalColumnAccessor(buffer: ByteBuffer, dataType: CalendarIntervalType)
+  extends BasicColumnAccessor[CalendarInterval](buffer, CALENDAR_INTERVAL)
+    with NullableColumnAccessor
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnAccessor.scala]: nit: two indents along with the others.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala [hunk 2]
```diff
@@ -295,6 +295,28 @@ private[columnar] final class BinaryColumnStats extends ColumnStats {
     Array[Any](null, null, nullCount, count, sizeInBytes)
 }
 
+private[columnar] final class IntervalColumnStats extends ColumnStats {
+  protected var upper: CalendarInterval =
+    new CalendarInterval(Int.MinValue, Int.MinValue, Long.MinValue)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala]: How about defining constant values for min/max, e.g., `CalendarInterval.MinValue`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala [hunk 3]
```diff
@@ -295,6 +295,28 @@ private[columnar] final class BinaryColumnStats extends ColumnStats {
     Array[Any](null, null, nullCount, count, sizeInBytes)
 }
 
+private[columnar] final class IntervalColumnStats extends ColumnStats {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala]: (This is not related to this pr though...) Can we already support column stats for intervals? Can you check?
https://github.com/apache/spark/blob/16f1b23d75c0b44aac61111bfb2ae9bb0f3fab68/sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala#L139
https://github.com/apache/spark/blob/16f1b23d75c0b44aac61111bfb2ae9bb0f3fab68/sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala#L238

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala [hunk 4]
```diff
@@ -705,6 +705,35 @@ private[columnar] case class MAP(dataType: MapType)
   override def clone(v: UnsafeMapData): UnsafeMapData = v.copy()
 }
 
+private[columnar] object CALENDAR_INTERVAL extends ColumnType[CalendarInterval]
+  with DirectCopyColumnType[CalendarInterval]{
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: nit: `]{` -> `] {`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/GenerateColumnAccessor.scala [hunk 5]
```diff
@@ -109,6 +115,14 @@ object GenerateColumnAccessor extends CodeGenerator[Seq[DataType], ColumnarItera
...
             }
            """
+        case CalendarIntervalType =>
+          // For CalendarInterval, it should have 16 bytes to store months(Int), days(Int),
+          // microseconds(Long) for future update even it's null now.
+          s"""
+            if (mutableRow.isNullAt($index)) {
+              rowWriter.write($index, (CalendarInterval) null);
+            }
+           """
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/GenerateColumnAccessor.scala]: Is this patch work related to this discussion? https://github.com/apache/spark/pull/26680#discussion_r351153420
If we can remove this patch, the gen'd code looks nicer to me. cc: @cloud-fan

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql [hunk 6]
```diff
@@ -264,3 +264,6 @@ select interval 'interval \t 1\tday';
 select interval 'interval\t1\tday';
 select interval '1\t' day;
 select interval '1 ' day;
+
+cache table interval_columnar as select i, cast(v as interval) from VALUES(1, '1 seconds'), (1, '2 seconds'), (2, NULL), (2, NULL) t(i,v);
+select * from interval_columnar;
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/interval.sql]: What's a purpose of this test?

**Review by maropu — COMMENTED** [review]
Can you also add end-to-end tests in `CachedTableSuite`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala]: this is nice

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala]: yes, these can be set to true for intervals

File: sql/core/src/test/resources/sql-tests/inputs/interval.sql (hunk 6 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/interval.sql]: maybe should be moved to `CachedTableSuite ` instead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala (hunk 3 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala]: Ah, I see. Thanks for the check. If you have time, do you wanna make a pr for that?

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala]: yea, I'd like to make a followup after this pull request, thanks for notice me this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/GenerateColumnAccessor.scala (hunk 5 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/GenerateColumnAccessor.scala]: After a second thought, I guess this patch is needless for CalendarInterval which is a `DirectCopyColumnType`

**Comment by yaooqinn** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnStatsSuite.scala [hunk 7]
```diff
@@ -30,6 +31,9 @@ class ColumnStatsSuite extends SparkFunSuite {
   testColumnStats(classOf[DoubleColumnStats], DOUBLE, Array(Double.MaxValue, Double.MinValue, 0))
   testColumnStats(classOf[StringColumnStats], STRING, Array(null, null, 0))
   testDecimalColumnStats(Array(null, null, 0))
+  testIntervalColumnStats(Array(
+    new CalendarInterval(Int.MaxValue, Int.MaxValue, Long.MaxValue),
+    new CalendarInterval(Int.MinValue, Int.MinValue, Long.MinValue), 0))
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnStatsSuite.scala]: nit Use `CalendarInterval.MIN_VALUE/MAX_VALUE` instead?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnTypeSuite.scala [hunk 8]
```diff
@@ -76,6 +78,8 @@ class ColumnTypeSuite extends SparkFunSuite with Logging {
     checkActualSize(ARRAY_TYPE, Array[Any](1), 4 + 8 + 8 + 8)
     checkActualSize(MAP_TYPE, Map(1 -> "a"), 4 + (8 + 8 + 8 + 8) + (8 + 8 + 8 + 8))
     checkActualSize(STRUCT_TYPE, Row("hello"), 28)
+    checkActualSize(CALENDAR_INTERVAL,
+      new CalendarInterval(Int.MaxValue, Int.MaxValue, Long.MaxValue), 16)
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnTypeSuite.scala]: nit: Use `CalendarInterval.MAX_VALUE` instead?

**Review by maropu — APPROVED** [review]
LGTM also cc: @viirya

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala [hunk 9]
```diff
@@ -705,6 +705,35 @@ private[columnar] case class MAP(dataType: MapType)
...
+  override def extract(buffer: ByteBuffer): CalendarInterval = {
+    ByteBufferHelper.getInt(buffer)
+    val months = ByteBufferHelper.getInt(buffer)
+    val days = ByteBufferHelper.getInt(buffer)
+    val microseconds = ByteBufferHelper.getLong(buffer)
+    new CalendarInterval(months, days, microseconds)
+  }
+
+  override def append(v: CalendarInterval, buffer: ByteBuffer): Unit = {
+    ByteBufferHelper.putInt(buffer, 16)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: if the length is fixed (16), can we skip it?

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: this is needed for `MutableUnsafeRow`

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: please check here https://github.com/apache/spark/blob/e1ea806b3075d279b5f08a29fe4c1ad6d3c4191a/sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala#L411

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: Or we can implement `extract` directly.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnarTestUtils.scala [hunk 10]
```diff
@@ -51,6 +51,8 @@ object ColumnarTestUtils {
       case DOUBLE => Random.nextDouble()
       case STRING => UTF8String.fromString(Random.nextString(Random.nextInt(32)))
       case BINARY => randomBytes(Random.nextInt(32))
+      case CALENDAR_INTERVAL =>
+        new CalendarInterval(Random.nextInt(), Random.nextInt(), 0)
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnarTestUtils.scala]: any reason to assign 0 here?

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnarTestUtils.scala]: nope, just for brevity, and the microseconds part is also covered by other tests.

**Comment by viirya** [comment]
Looks fine to me. cc @cloud-fan

**Comment by maropu** [comment]
Thanks for the check, @viirya

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by yaooqinn** [comment]
Thanks @dongjoon-hyun for bringing up jenkins.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala [hunk 11]
```diff
@@ -705,6 +705,35 @@ private[columnar] case class MAP(dataType: MapType)
   override def clone(v: UnsafeMapData): UnsafeMapData = v.copy()
 }
 
+private[columnar] object CALENDAR_INTERVAL extends ColumnType[CalendarInterval]
+  with DirectCopyColumnType[CalendarInterval] {
+
+  override def dataType: DataType = CalendarIntervalType
+
+  override def defaultSize: Int = 16
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: We need to override `actualSize`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala [hunk 12]
```diff
@@ -295,6 +295,26 @@ private[columnar] final class BinaryColumnStats extends ColumnStats {
...
+  protected var upper: CalendarInterval = CalendarInterval.MIN_VALUE
+  protected var lower: CalendarInterval = CalendarInterval.MAX_VALUE
+
+  override def gatherStats(row: InternalRow, ordinal: Int): Unit = {
+    if (!row.isNullAt(ordinal)) {
+      val value = row.getInterval(ordinal)
+      if (value.compareTo(upper) > 0) upper = value
+      if (value.compareTo(lower) < 0) lower = value
+      sizeInBytes += CALENDAR_INTERVAL.defaultSize
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnStats.scala]: Shall we use `actualSize` here?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnTypeSuite.scala [hunk 13]
```diff
@@ -76,6 +78,7 @@ class ColumnTypeSuite extends SparkFunSuite with Logging {
     checkActualSize(ARRAY_TYPE, Array[Any](1), 4 + 8 + 8 + 8)
     checkActualSize(MAP_TYPE, Map(1 -> "a"), 4 + (8 + 8 + 8 + 8) + (8 + 8 + 8 + 8))
     checkActualSize(STRUCT_TYPE, Row("hello"), 28)
+    checkActualSize(CALENDAR_INTERVAL, CalendarInterval.MAX_VALUE, 16)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnTypeSuite.scala]: `16` -> `20` (or `4 + 4 + 4 + 8`)?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnarTestUtils.scala (hunk 10 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/columnar/ColumnarTestUtils.scala]: Shall we use `Random.nextLong()` for consistency?
Also, cc @gengliangwang since this is random data generation.

File: sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala [hunk 14]
```diff
@@ -1094,4 +1096,18 @@ class CachedTableSuite extends QueryTest with SQLTestUtils with SharedSparkSessi
...
+      Seq((1, "1 second"), (2, "2 seconds"), (2, null))
+        .toDF("k", "v").write.format("json").saveAsTable("interval_cache")
+      sql("CACHE TABLE t1 AS SELECT k, cast(v as interval) FROM interval_cache")
+      assert(spark.catalog.isCached("t1"))
+      checkAnswer(sql("SELECT * FROM t1 WHERE k = 1"),
+        Row(1, new CalendarInterval(0, 0, DateTimeConstants.MICROS_PER_SECOND)))
+      sql("UNCACHE TABLE t1")
+      assert(!spark.catalog.isCached("t1"))
+    }
+
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala]: nit. Redundant empty line.

File: sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala [hunk 15]
```diff
@@ -1094,4 +1096,18 @@ class CachedTableSuite extends QueryTest with SQLTestUtils with SharedSparkSessi
       }
     }
   }
+
+  test("cache supports for intervals") {
+    withTable("interval_cache") {
+      Seq((1, "1 second"), (2, "2 seconds"), (2, null))
+        .toDF("k", "v").write.format("json").saveAsTable("interval_cache")
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala]: Shall we remove `.format("json")`?

**Review by dongjoon-hyun — COMMENTED** [review]
I left a few comments. Could you update this PR once more, @yaooqinn ? Thanks!

**Comment by yaooqinn** [comment]
Thanks very much for your comments @dongjoon-hyun

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @yaooqinn , @maropu , @viirya .
Merged to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala [hunk 16]
```diff
@@ -705,6 +705,37 @@ private[columnar] case class MAP(dataType: MapType)
...
+
+  override def actualSize(row: InternalRow, ordinal: Int): Int = 20
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: why actual size is different from default size?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala [hunk 17]
```diff
@@ -705,6 +705,37 @@ private[columnar] case class MAP(dataType: MapType)
...
+
+  override def getField(row: InternalRow, ordinal: Int): CalendarInterval = row.getInterval(ordinal)
+
+  override def setField(row: InternalRow, ordinal: Int, value: CalendarInterval): Unit = {
+    row.setInterval(ordinal, value)
+  }
+
+  override def extract(buffer: ByteBuffer): CalendarInterval = {
+    ByteBufferHelper.getInt(buffer)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: the first int is the size?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: Now I got why the actual size is 4 bytes larget than default size. We should document it.

BTW can we skip storing the size? It's fixed anyway.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala]: yes, we can skip it, the current implementation is just to reuse copy logic of `MutableUnsafeRow`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala [hunk 18]
```diff
@@ -719,6 +750,7 @@ private[columnar] object ColumnType {
       case DoubleType => DOUBLE
       case StringType => STRING
       case BinaryType => BINARY
+      case i: CalendarIntervalType => CALENDAR_INTERVAL
```
Comment (Peng-Lei) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala:753]: "i" is no need
